### PR TITLE
Adjustement to transform another dataset along side the data without learning the embedding on this secondary dataset

### PR DESCRIPTION
**Pull request by WiestDaessle**
A way to train on a dataset and apply to another dataset at the same time

### CODE REVIEW

**Comment by hyhuang00** [comment]
Thank you again for your help! I have reviewed the code and I would try to provide a summarization of my understanding here:

- An extra parameter Xp can be provided in the `fit()` and `fit_transform()` function that holds the test data.
- The test data will only be attached towards the original data via nearest neighbor pairs.
- During the pacmap updates, the test data pairs will only provide a gradient to themselves, therefore the training data is unaffected.

Am I understanding it correctly? Seems like the mid-near pairs and further pairs for the test data are missing, but my preliminary test shows that the embedding is still performing well. Please let me know if I've made any mistakes in understanding your code!

It seems like currently there are some conflicts that needs to be resolved due to the newly introduced support on different distance metrics and I will work on resolving them right away.

**Comment by WiestDaessle** [comment]
Indeed,

1) i added an input for a secondary dataset X', which we want the embedding
to be applied on, but we do not want this data to affect the embedding
2) Yes, the idea is to transform this dataset, so we'd like to only apply
the nearest neighbor transform, as we would not really know what to apply
for this data for MD&FP points, as this secondary dataset may also lie in
unknown area of the original space, where the embedding has not really
learned anything
3) Yes the gradient are only modifying X' coordinates, and the presence of
the X' data is not modifying any of the X gradients,

Indeed  midnear & further points aren't taken into account, but it's on
purpose, as the idea is to adjust the data with the learned embedding, so
applying closest transform, i discarded the mid & far points has they
didn't sound like a good idea to adjust (the idea could be like feeding
X&X' with the same dataset, the results should be a perfect projection
(well we may need to adjust the number of neighborhs for X' (and set to it
to 1) for perfect match, but it's more a toy example then and not exactly
what we expect in true datasets (btw i tested with  X' == X and with the
closest neighbor).

So basically you got all the points correctly :D
Also i was thinking about setting another n_neighbor parameter for the X',
but even if it gives more control to the user, it may need the users to get
a deeper understanding on what's going on, but this may be a good option.

Le lun. 21 févr. 2022 à 08:58, Haiyang Huang ***@***.***> a
écrit :
